File: CS-IOS/Classes/Core/Classes/Controller/CSSearchBarController.m
//
// Created by Rene Dohan on 21/01/15.
// Copyright (c) 2015 creative_studio. All rights reserved.
//

#import "CSSearchBarController.h"

@implementation CSSearchBarController {
    UIViewController *_parent;
    BOOL _searchBarShouldBeginEditing;
    void (^_onTextDidChange)(NSString *);
}

- (instancetype)construct:(CSMainController *)parent :(UISearchBar *)bar :(void (^)(NSString *))onTextDidChange {
    [super construct:parent];
    _parent = parent;
    _searchBar = bar;
    _onTextDidChange = [onTextDidChange copy];
    _searchBar.delegate = self;
    return self;
}

- (instancetype)construct:(CSMainController *)parent :(void (^)(NSString *))onTextDidChange {
    [self construct:parent :UISearchBar.createEmpty :onTextDidChange];
    return self;
}

- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText {
    if (![searchBar isFirstResponder]) {
        _searchBarShouldBeginEditing = NO;
        [searchBar resignFirstResponder];
    }
    runWith(_onTextDidChange, searchText);
}

- (void)searchBarCancelButtonClicked:(UISearchBar *)searchBar {
    [searchBar resignFirstResponder];
}

- (void)searchBarSearchButtonClicked:(UISearchBar *)searchBar {
    [searchBar resignFirstResponder];
}

- (BOOL)searchBarShouldBeginEditing:(UISearchBar *)bar {
    BOOL boolToReturn = _searchBarShouldBeginEditing;
    _searchBarShouldBeginEditing = YES;
    return boolToReturn;
}

- (NSString *)text {
    return _searchBar.text;
}
@end

File: Renetik/Renetik/Classes/Core/Classes/Controller/CSSearchBarController.swift
//
// Created by Rene Dohan on 12/17/19.
//

import RenetikObjc
import UIKit

public class CSSearchBarController: CSMainController, UISearchBarDelegate {
    public let bar = UISearchBar.construct().resizeToFit()
    public var text: String {
        bar.text ?? ""
    }

    private var searchBarShouldBeginEditing = false
    private var onTextChanged: ((String) -> Void)!

    @discardableResult
    public func construct(by parent: CSMainController,
                          placeHolder: String = .searchPlaceholder,
                          onTextChanged: @escaping (String) -> Void) -> Self {
        super.constructAsViewLess(in: parent)
        self.onTextChanged = onTextChanged
        bar.delegate = self
        bar.showsCancelButton = false
        bar.placeholder = placeHolder
        return self
    }

    public func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
        if !searchBar.isFirstResponder {
            searchBarShouldBeginEditing = false
            searchBar.resignFirstResponder()
        }
        onTextChanged(searchText)
    }

    public func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
        searchBar.resignFirstResponder()
    }

    public func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
        searchBar.resignFirstResponder()
    }

    public func searchBarShouldBeginEditing(_: UISearchBar) -> Bool {
        let boolToReturn = searchBarShouldBeginEditing
        searchBarShouldBeginEditing = true
        return boolToReturn
    }

    public func searchBarTextDidBeginEditing(_ searchBar: UISearchBar) {
        animate { searchBar.showsCancelButton = true }
    }

    public func searchBarTextDidEndEditing(_ searchBar: UISearchBar) {
        animate { searchBar.showsCancelButton = false }
    }
}

/// CSSearchBarController+CSTableController
public extension CSSearchBarController {
    @discardableResult
    func construct<Row: CSTableControllerRow & CustomStringConvertible, Data>(
        _ parent: CSMainController,
        placeHolder: String = .searchPlaceholder,
        table: CSTableController<Row, Data>
    ) -> Self {
        let tableFilter = TableFilter<Row, Data>()
        table.filter = tableFilter
        construct(by: parent, placeHolder: placeHolder) { string in
            tableFilter.searchText = string
            table.filterDataAndReload()
        }
        return self
    }
}

private class TableFilter<Row: CSTableControllerRow & CustomStringConvertible, Data>: CSTableControllerFilter<Row, Data>
{
    var searchText = ""

    override func filter(data: [Row]) -> [Row] {
        data.filter(bySearch: searchText)
    }
}
